Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java

```diff
@@ -220,7 +220,100 @@ protected boolean leftOf(float[] point, short nodeOffset) {
+                    : internalNodes.mass[siblingOffset];
+            short leafOffset = (short) (leafNodes.add(NULL, pointIndex, 1) + maxSize); // parent pointIndex
+            // needs to be fixed
+            short mergedNode = leftOf(point, cutDimension, cutValue)
+                    ? internalNodes.addNode(NULL, leafOffset, siblingOffset, cutDimension, cutValue, oldmass + 1)
+                    : internalNodes.addNode(NULL, siblingOffset, leafOffset, cutDimension, cutValue, oldmass + 1);
+
+            int parent;
+            if (isLeaf(siblingOffset))
+                parent = leafNodes.parentIndex[siblingOffset - maxSize];
+            else
+                parent = internalNodes.parentIndex[siblingOffset];
+
+            if (parent == NULL) {
+                rootIndex = mergedNode;
+            } else {
+                replaceNode(siblingOffset, mergedNode);
+            }
+            leafNodes.parentIndex[leafOffset - maxSize] = mergedNode;
+            if (isLeaf(siblingOffset))
+                leafNodes.parentIndex[siblingOffset - maxSize] = mergedNode;
+            else
+                internalNodes.parentIndex[siblingOffset] = mergedNode;
+
+            internalNodes.boundingBox[mergedNode] = savedBox;
+            short tempNode = mergedNode;
+            while (internalNodes.parentIndex[tempNode] != parentIndex) {
+                tempNode = internalNodes.parentIndex[tempNode];
+                internalNodes.boundingBox[tempNode].addPoint(point);
+            }
+
+        }
+
+        boolean checkContainsAndUpdateCut(double[] point, short nodeOffset) {
+            BoundingBox existingBox = getBoundingBox(nodeOffset);
+            if (existingBox.contains(point)) {
+                return true;
+            }
+            BoundingBox mergedBox = existingBox.getMergedBox(point);
+            Cut cut = randomCut(random, mergedBox);
+            int splitDimension = cut.getDimension();
+            double splitValue = cut.getValue();
+            double minValue = existingBox.getMinValue(splitDimension);
+            double maxValue = existingBox.getMaxValue(splitDimension);
+
+            // if the proposed cut separates the new point from the existing bounding box:
+            // * create a new leaf node for the point
+            // * make it a sibling of the existing bounding box
+            // * make the new leaf node and the existing node children of a new node with
+            // the merged bounding box
+
+            if (minValue > splitValue || maxValue <= splitValue) {
+                cutValue = splitValue;
+                cutDimension = splitDimension;
+                siblingOffset = nodeOffset;
+                savedBox = mergedBox;
+            }
+            return false;
+        }
+    }
+
+    CandidateChange candidateChange = new CandidateChange();
+
+    boolean checkEquals(double[] point, double[] otherPoint) {
```

Comment:
minor. inputs should be sanitized and array comparison can be done by Arrays.equals without an implementation everywhere it is needed.

Reply:
Agree. This is a discussion point - we now have equals in the PointStore. But the issue is that the equals in the addPoint has to be in sync with the randomCut function. Maybe we remove the pointEquals in PointStore?